src/query/storage/m3/storage.go

```diff
@@ -446,7 +450,8 @@ func (s *m3storage) fetchCompressed(
 
 			session := namespace.Session()
 			namespaceID := namespace.NamespaceID()
-			iters, metadata, err := session.FetchTagged(ctx, namespaceID, m3query, queryOptions)
+			narrowedQueryOpts := narrowQueryOpts(queryOptions, namespace)
+			iters, metadata, err := session.FetchTagged(ctx, namespaceID, m3query, narrowedQueryOpts)
```

Comment:
So I was thinking. User passes start,end for query. We invoke resolveClusterNamespacesForQuery that potentially can override this range. Currently we call it `narrowing`. Maybe we can be more generic. Say after we call `resolveClusterNamespacesForQuery` we always get namespaces with "actual range set". Sometimes it matches the user specified start,end, sometimes it's overriden adjusting based on data latency.

Reply:
See above.